src/Network/Transport/TCP.hs

```diff
@@ -910,6 +913,23 @@ handleConnectionRequest transport socketClosed sock = handle handleException $ d
       let maxAddressLength = tcpMaxAddressLength $ transportParams transport
       theirAddress <- EndPointAddress . BS.concat <$>
         recvWithLength maxAddressLength sock
+      (theirHost, theirPort, theirEndPointId)
+        <- maybe (throwIO (userError "handleConnectionRequest: peer gave malformed address"))
+                 return
+                 (decodeEndPointAddress theirAddress)
+      -- If the OS-determined host doesn't match the host that the peer gave us,
+      -- then we have no choice but to reject the connection. It's because we
+      -- use the EndPointAddress to key the remote end points (localConnections)
+      -- and we don't want to allow a peer to deny service to other peers by
+      -- claiming to have their host and port.
+      unless (theirHost == actualHost) $ do
+        let errorString = concat [
+                "handleConnectionRequest: reported host "
+              , show theirHost
+              , " does not match actual host "
+              , show actualHost
+              ]
```

Comment:
I'm just realizing the error message is not visible to the user because `handleException` discards the exception.